[Security Solution] Install/Update Prebuilt Rules Test Implementation refactor

jpdjere opened the pull request:
Fixes: https://github.com/elastic/kibana/issues/148192

(Tick the two open checkboxes in that issue when merging this PR)

## Summary

This PR rewrites/refactors Cypress tests for the Installation and Upgrade of Prebuilt Rules implemented in https://github.com/elastic/kibana/pull/161687. Most of the changes here address feedback received in that PR - answered those comments there.

- RBAC/Authorization: adds tests scenarios for users with full privileges (happy path)
- Gets rid of huge util helpers such as `assertRuleAvailableForInstallAndInstallOne` and rewrites test cases in a more descriptive way, with step by step actions.
- Gets rid of complex logic in tests and their helpers - removing if/else logic within them and removing optional flags passed to helpers.
- Fixes `bulkCreateRuleAssets` util and uses it in other helpers to install multiple `security-rule` assets with a single bulk request to ES.

Additionally: checked `installation_and_upgrade.md` test plan to make sure it matches with the test in place. Added [link](https://github.com/elastic/kibana/issues/166215) to a ticket for a to-do task for the sections:
- Rule installation workflow: filtering, sorting, pagination
- Rule upgrade workflow: filtering, sorting, pagination

## Flaky test runner

~~https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds/3420~~ ~~🟢~~ 

https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds/3513

<!--ONMERGE {"backportTargets":["8.11"]} ONMERGE-->

elasticmachine commented:
Pinging @elastic/security-detections-response (Team:Detections and Resp)

elasticmachine commented:
Pinging @elastic/security-solution (Team: SecuritySolution)

xcrzx commented on x-pack/test/security_solution_cypress/cypress/tasks/prebuilt_rules.ts:
```diff
@@ -35,190 +29,93 @@ export const ruleUpdatesTabClick = () => {
-      cy.get(rule['security-rule'].name).should('not.exist');
-    }
+
+  for (const rule of rules) {
+    cy.get(RULES_UPDATES_TABLE).contains(rule['security-rule'].name);
```
Why do we need this check here?

xcrzx commented on x-pack/test/security_solution_cypress/cypress/tasks/prebuilt_rules.ts:
```diff
@@ -35,190 +29,93 @@ export const ruleUpdatesTabClick = () => {
+  const toastMessage = `${rules.length} ${rulesString} updated successfully.`;
+  cy.get(TOASTER).should('be.visible').should('have.text', toastMessage);
   for (const rule of rules) {
-    cy.get(RULES_UPDATES_TABLE).contains(rule['security-rule'].name);
+    cy.get(rule['security-rule'].name).should('not.exist');
   }
```
Not sure I understand what this check does. Could you please add a comment?

xcrzx commented on x-pack/test/security_solution_cypress/cypress/tasks/prebuilt_rules.ts:
```diff
@@ -35,190 +29,93 @@ export const ruleUpdatesTabClick = () => {
+    );
+  }
+  for (const rule of rules) {
+    cy.get(getInstallSingleRuleLoadingSpinnerByRuleId(rule['security-rule'].rule_id)).should(
```
Looks like a typo, we need to check the "upgrade" spinner here

jpdjere replied:
Thanks for catching this! Fixing

xcrzx reviewed (commented):
Thanks for refactoring the tests, @jpdjere! They're much more readable now 👍 I've left a few comments for additional improvements, but they're pretty minor. Reach out if we need to discuss anything.

banderror commented:
@xcrzx Do you think this PR is a great improvement compared to the previous implementation? Would we benefit from merging it now and addressing the comments you left in a separate PR? I'm asking this because @jpdjere won't be able to work on this PR until mid-October when he gets back from leave, and any improvements to tests' quality and stability we can merge now will contribute to our 8.11 goals.

xcrzx commented:
> @xcrzx Do you think this PR is a great improvement compared to the previous implementation? Would we benefit from merging it now and addressing the comments you left in a separate PR? I'm asking this because @jpdjere won't be able to work on this PR until mid-October when he gets back from leave, and any improvements to tests' quality and stability we can merge now will contribute to our 8.11 goals.

I think it's worth merging this PR. I would only run it against the flaky runner to ensure we do not introduce any flakiness.

jpdjere commented on x-pack/test/security_solution_cypress/cypress/e2e/detection_response/prebuilt_rules/prebuilt_rules_install_update_error_handling.cy.ts:
```diff
@@ -45,37 +56,48 @@ describe(
+        // Make two mock rules available for installation
+        installPrebuiltRuleAssets([RULE_1, RULE_2]);
+        // Navigate to install Elastic rules page
+        addElasticRulesButtonClick();
```
Done :+1

jpdjere commented on x-pack/test/security_solution_cypress/cypress/e2e/detection_response/prebuilt_rules/prebuilt_rules_install_update_authorization.cy.ts:
```diff
@@ -120,11 +115,65 @@ describe(
+
+    describe('User with full privileges on Security Solution', () => {
+      beforeEach(() => {
+        preventPrebuiltRulesPackageInstallation();
```
Removed

jpdjere commented on x-pack/test/security_solution_cypress/cypress/e2e/detection_response/prebuilt_rules/prebuilt_rules_install_update_authorization.cy.ts:
```diff
@@ -120,11 +115,65 @@ describe(
+    describe('User with full privileges on Security Solution', () => {
+      beforeEach(() => {
+        preventPrebuiltRulesPackageInstallation();
+        loginPageAsFullPrivilegesUser();
```
Yes, makes sense. I'm using now the `Hunter/ T3 Analyst` role, following this table: https://github.com/elastic/kibana/pull/81866#issue-511165754

jpdjere replied:
Looks like the role with the minimum required priviliges for writing rules

jpdjere commented on x-pack/test/security_solution_cypress/cypress/e2e/detection_response/prebuilt_rules/prebuilt_rules_install_update_authorization.cy.ts:
```diff
@@ -39,48 +46,48 @@ const UPDATED_RULE_1 = createRuleAssetSavedObject({
-        rule_id: 'rule_2',
-      });
       beforeEach(() => {
+        preventPrebuiltRulesPackageInstallation();
```
Removing

jpdjere commented on x-pack/test/security_solution_cypress/cypress/e2e/detection_response/prebuilt_rules/prebuilt_rules_install_update_authorization.cy.ts:
```diff
@@ -39,48 +46,48 @@ const UPDATED_RULE_1 = createRuleAssetSavedObject({
+      preventPrebuiltRulesPackageInstallation();
+      cleanKibana();
+
+      // Install one prebuilt rule asset to assert that user can't install it
+      installPrebuiltRuleAssets([RULE_2]);
+
+      // Install one prebuilt rule asset to assert that user can't upgrade it
+      createAndInstallMockedPrebuiltRules([OUTDATED_RULE_1]);
+      // Create a new version of the rule to make it available for upgrade
+      installPrebuiltRuleAssets([UPDATED_RULE_1]);
```
Fixing

jpdjere commented:
@xcrzx @banderror Thanks for the review. I rebased, updated and addressed the additional feedback. I'll run the flaky test runner and will merge if everything is green.

banderror commented:
@jpdjere Please be careful with these changes since tests for prebuilt rules have been unskipped and enabled in Serverless. We need two flaky test runner jobs for this PR: one for ESS and another one for Serverless.

jpdjere commented:
@banderror Yes, the flaky test runner I pasted above includes both tasks (ESS and serverless). I'll check what happened in the last CI run and fix before merging, but both tasks passed with green.

jpdjere commented:
@elasticmachine merge upstream

kibanamachine commented:
merge conflict between base and head

banderror commented:
@jpdjere It doesn't look like any tests have run though, if you check the logs. The spec pattern `./cypress/e2e/!(investigations\|explore)/**/detection_rules/**/*.cy.ts` is incorrect:

```
info Config spec pattern: ./cypress/e2e/**/*.cy.ts
--
  | info Arguments spec pattern: ./cypress/e2e/!(investigations\|explore)/**/detection_rules/**/*.cy.ts
  | info Resulting spec pattern: ./cypress/e2e/!(investigations\|explore)/**/detection_rules/**/*.cy.ts
  | @cypress/grep: filtering using tag(s) "@serverless --@brokenInServerless --@skipInServerless"
  | @cypress/grep: will omit filtered tests
  | grep and/or grepTags has eliminated all specs
  | grepTags: @serverless --@brokenInServerless --@skipInServerless
  | Will leave all specs to run to filter at run-time
  | info Resolved spec files or pattern after grep: ./cypress/e2e/!(investigations\|explore)/**/detection_rules/**/*.cy.ts
  | info No tests found - all tests could have been skipped via Cypress tags


```

kibanamachine commented:
## 💔 All backports failed

| Status | Branch | Result |
|:------:|:------:|:------|
|❌|8.11|**Backport failed because of merge conflicts**<br><br>You might need to backport the following PRs to 8.11:<br> - [[Security Solution] Adding serverlessQA tag (#167494)](https://github.com/elastic/kibana/pull/167494)|

### Manual backport
To create the backport manually run:
```
node scripts/backport --pr 165488
```

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sqren/backport)

<!--- Backport version: 8.9.7 -->
